stdlib/public/core/Substring.swift.gyb

```diff
@@ -19,7 +19,17 @@ extension String {
   ///
   /// - Complexity: O(*n*), where *n* is the length of `substring`.
   public init(_ substring: Substring) {
-    self = String(substring._slice)
+    let x = substring._wholeString
+    let start = substring.startIndex
+    let end = substring.endIndex
+    let u16 = x._core[start.encodedOffset..<end.encodedOffset]
+    if start.samePosition(in: x.unicodeScalars) != nil
+    && end.samePosition(in: x.unicodeScalars) != nil {
+      self = String(_StringCore(u16))
```

Comment:
Thanks, that makes a little more sense. I'm unlikely to have time, but very interested in building up my expertise here.

Out-of-scope aside: The fact that all non-NSString Strings always have valid UTF16 seems a little surprising, as the subscript for UTF16View does surrogate pair validation on every access, not just when the cocoaString property is set. This results in a ~50% slowdown for all-valid Strings compared to operating directly on the _StringCore. Maybe this is because the held _StringCore might be in slice mode, and perhaps that doesn't enforce the validity currently. Food for thought for a new _StringCore.